# Improving *Corylus avellana* disease management: using metagenomics to illuminate the rhizosphere microbiome of *Corylus*

This repository contains the scripts and input files required to repeat the analyses reported in the associated article.

## 1_read_quality_control

This folder contains the scripts used to clean the Illumina reads prior to assembly.

## 2_metagenome_assembly

This folder contains the scripts used to assemble the metagenomes.

## 3_metagenome_binning

This folder contains the scripts used to bin the metagenomes and identify high quality MAGs.

## 4_metagenome_and_mag_annotation

This folder contains the scripts used to annotate the metagenomes and MAGs, both in terms of taxonomic classification and functional annotation.

## 5_read_subsetting_and_reassembly

This folder contains the scripts used to subset the metagenomes to an equal read count, reassemble the metagenomes, and calculate stats.
